### PR TITLE
Log4j changes in java SDK

### PR DESCRIPTION
**Pull request by azmigproject**

### CODE REVIEW

File: device/iothub-java-client/src/main/java/com/microsoft/azure/iothub/CustomLogger.java [hunk 1]
```diff
@@ -0,0 +1,37 @@
...
+
+    public CustomLogger(Class<?> clazz)
+    {
+        logger = Logger.getLogger(clazz);
+    }
+
+    public void LogInfo(String message, Object...params)
+    {
+        if(logger.isEnabledFor(Level.INFO))
+        {
```
Comment (prmathur-microsoft) [inline device/iothub-java-client/src/main/java/com/microsoft/azure/iothub/CustomLogger.java]: Would it be possible to add granularity for each of these messages ? specifically - ALL, debug, info, fine, severe, off, trace, warn ?

File: device/iothub-java-client/src/main/java/com/microsoft/azure/iothub/transport/amqps/AmqpsIotHubConnection.java [hunk 2]
```diff
@@ -152,7 +154,7 @@ public AmqpsIotHubConnection(DeviceClientConfig config, Boolean useWebSockets) t
         // endpoint private member variables using the send/receiveEndpointFormat constants and device id.]
         this.sendEndpoint = String.format(sendEndpointFormat, deviceId);
         this.receiveEndpoint = String.format(receiveEndpointFormat, deviceId);
-
+        this.logger = new CustomLogger(AmqpsIotHubConnection.class);
```
Comment (prmathur-microsoft) [inline device/iothub-java-client/src/main/java/com/microsoft/azure/iothub/transport/amqps/AmqpsIotHubConnection.java]: Can you use get-class and avoid hard coding class names ?

File: device/iothub-java-client/src/main/java/com/microsoft/azure/iothub/transport/amqps/AmqpsIotHubConnection.java [hunk 3]
```diff
@@ -503,6 +518,7 @@ public void onReactorFinal(Event event)
                 openAsync();
             } catch (IOException e)
             {
+                logger.LogError(e);
                 e.printStackTrace();
             }
         }
```
Comment (prmathur-microsoft) [inline device/iothub-java-client/src/main/java/com/microsoft/azure/iothub/transport/amqps/AmqpsIotHubConnection.java]: Could add appropriate logging in each of these override methods using debug level ? At the very least provide a succesful entry logging.

File: device/iothub-java-client/src/main/java/com/microsoft/azure/iothub/transport/amqps/AmqpsIotHubConnection.java [hunk 4]
```diff
@@ -842,6 +861,10 @@ private String  convertToPem(String derPath) {
         return derPath + ".pem" ;
     }
 
+    private String getMethodName()
+    {
+        return Thread.currentThread().getStackTrace()[2].getMethodName();
+    }
```
Comment (prmathur-microsoft) [inline device/iothub-java-client/src/main/java/com/microsoft/azure/iothub/transport/amqps/AmqpsIotHubConnection.java]: Please avoid hard coding any numbers and define it for the class.

File: device/iothub-java-client/src/main/java/com/microsoft/azure/iothub/transport/amqps/AmqpsTransport.java [hunk 5]
```diff
@@ -74,6 +78,7 @@ public AmqpsTransport(final DeviceClientConfig config, Boolean useWebSockets)
 
         // Codes_SRS_AMQPSTRANSPORT_15_002: [The constructor shall set the transport state to CLOSED.]
         this.state = State.CLOSED;
+        this.logger = new CustomLogger(AmqpsTransport.class);
     }
```
Comment (prmathur-microsoft) [inline device/iothub-java-client/src/main/java/com/microsoft/azure/iothub/transport/amqps/AmqpsTransport.java]: Same here. use this.getClass

File: device/iothub-java-client/src/main/java/com/microsoft/azure/iothub/transport/amqps/AmqpsTransport.java [hunk 6]
```diff
@@ -106,6 +111,7 @@ public synchronized void open() throws IOException
```
Comment (prmathur-microsoft) [inline device/iothub-java-client/src/main/java/com/microsoft/azure/iothub/transport/amqps/AmqpsTransport.java]: Please make sure you log every exception eg I see it missed in #104

File: device/iothub-java-client/src/main/java/com/microsoft/azure/iothub/transport/amqps/AmqpsTransport.java [hunk 7]
```diff
@@ -168,13 +178,14 @@ public void sendMessages() throws IOException, IllegalStateException
...
         {
+            logger.LogError("Cannot send messages when the AMQPS transport is closed, method name is %s ", getMethodName());
             throw new IllegalStateException("Cannot send messages when the AMQPS transport is closed.");
         }
 
         // Codes_SRS_AMQPSTRANSPORT_15_013: [If there are no messages in the waiting list, the function shall return.]
         if (this.waitingMessages.size() <= 0)
         {
-            return;
+			return;
```
Comment (prmathur-microsoft) [inline device/iothub-java-client/src/main/java/com/microsoft/azure/iothub/transport/amqps/AmqpsTransport.java]: Please remove extra tab.

File: device/iothub-java-client/src/main/java/com/microsoft/azure/iothub/transport/amqps/AmqpsTransport.java [hunk 8]
```diff
@@ -201,7 +212,8 @@ public void sendMessages() throws IOException, IllegalStateException
                     // Codes_SRS_AMQPSTRANSPORT_15_036: [The function shall create a new Proton message from the IoTHub message.]
```
Comment (prmathur-microsoft) [inline device/iothub-java-client/src/main/java/com/microsoft/azure/iothub/transport/amqps/AmqpsTransport.java]: Please log that message has expired, sent, send successfull, receive along with their timestamps where ever appropriate in this file.

File: device/iothub-java-client/src/main/resources/log4j.properties [hunk 9]
```diff
@@ -0,0 +1,13 @@
+log4j.rootLogger = ALL, FILE
+
+log4j.appender.FILE=org.apache.log4j.RollingFileAppender
+
+log4j.appender.FILE.File=/home/samsung/Details.log
+
```
Comment (prmathur-microsoft) [inline device/iothub-java-client/src/main/resources/log4j.properties]: Do not hard code this path. Use a default location eg - current dir ?

**Comment by prmathur-microsoft** [comment]
Can you please address my comments and rebase your changes to the latest in master ?

**Comment by prmathur-microsoft** [comment]
Also forgot to mention to please add requirement docs, tests and verify all existing test pass.with the change.

**Comment by tameraw** [comment]
@azmigproject - Saw your new pull request, so closing this one and will review and merge the new PR.
https://github.com/Azure/azure-iot-sdk-java/pull/24

**Pull request closed without merging**